README.md
### 该仓库为 Marrs 组件库使用的示例仓库，当前开发预览版，问题比较多，部分 api 可能不稳定，有任何问题请在企业微信群中提出，如未得到及时反馈请单独联系 柴必青

#### 部分组件已经完成了类型文件编写，关于组件 api 可以参考对应文件或者示例中的用法。

#### 渐进式的心智负担

- 普通使用
  1. react
  2. html/css/jsx
- 高级使用
  1. sx 工具
  2. classes-slot
  3. theme 修改: 全局 props、组件配置、调色盘等，生成自己的主题
- 进阶使用
  1. theme 在组件外代码的使用
  2. jss
  3. styled component
  4. 使用 makeStyles 生成样式或结合 classes-slot 修改组件库默认样式

#### 对于 styled component 概念或者 jss 有不理解或者困惑的请及时联系提出

#### 以下复制自组件库仓库 README.md

### 当前遗留问题或未完善解决问题

- [ ] 文档 api 站点
- [ ] ts 类型文件补齐
- [ ] 词法优化
- [ ] 性能优化
- [ ] 语义优化
- [ ] 注释补齐
- [ ] SFP 补齐
- [ ] 暴露 classname 补齐
- [ ] 参数与 api 稳定性
- [ ] 一期完整的组件
- [ ] 测试脚本

## Alpha 开发预览版

```shell
yarn add @wemo-ui/marrs@alpha
// 或者
npm i @wemo-ui/marrs@alpha
```

# 概念介绍

## 主题 Theme

Marrs 拥有一套默认的主题，该默认主题提取自微盟设计团队的设计原稿。开发者也可以根据自身项目二次调整主题配置，生成属于自己的自定义主题

默认主题不需要特殊配置默认生效，自定义主题有些简单步骤

### 自定义主题的使用

Marrs 支持`运行时修改主题`&`主题嵌套与隔离`。下面是个极端例子

```react
import { Button } from '@wemo-ui/marrs';
import { createTheme, ThemeProvider } from '@wemo-ui/marrs/styles';

const myTheme = createTheme({
  palette: {
    primary: { main: '#00ff00' }
  }
});

const nestTheme = createTheme({
  palette: {
    primary: { main: '#ff0000' }
  }
});

const cusTheme = createTheme({
  components: {
    MarrsButton: {
      defaultProps: {
        volume: 'tiny',
        color: 'info',
        textTransform: 'uppercase'
      }
    }
  }
});

export default function CartView() {
  return (
    <div style={{ padding: 100 }}>
      <ThemeProvider theme={myTheme}>
        <Button>myTheme</Button>
        <ThemeProvider theme={nestTheme}>
          <Button>nestTheme</Button>
        </ThemeProvider>
        <ThemeProvider theme={cusTheme}>
          <Button>cusTheme</Button>
          <Button color="success" volume="medium">
            cusTheme2
          </Button>
        </ThemeProvider>
      </ThemeProvider>
    </div>
  );
}

```

![image-20210818155252442](https://image-c.weimobwmc.com/ol-6LEqi/74860ac20b46478d9d1ee846394e715c.png)

### 主题的制作

主题配置详解

```javascript
{
  // 调色盘
  palette: {
    // [warning]以下内容必须存在，可以修改但不可置空，否则有可能会影响组件库的运行
    common: {
      black: '#000000',
      white: '#ffffff'
    },
    // 快速模式切换, alpha
    mode: 'main',

    // 色调调色
    primary: {
      main: '#FA2C19'
    },
    secondary: {
      main: '#9C27B0'
    },
    error: {
      main: '#FF2E2E'
    },
    warning: {
      main: '#FFA300',
      // contrastText代表对应main为背景色时候的对照字体颜色，如果没有显示声明，默认补全为getContrastText(colorValue)
      contrastText: 'rgba(0, 0, 0, 0.87)'
    },
    info: {
      main: '#2A6AE9'
    },
    success: {
      main: '#07C160'
    },
    // 中性色阶
    neutral: {
      N1: '#FFFFFF',
      N2: '#FAFAFA',
      N3: '#F5F5F5',
      N4: '#F2F2F2',
      N5: '#E0E0E0',
      N6: '#C4C4C4',
      N7: '#9E9E9E',
      N8: '#757575',
      N9: '#212121'
    },
    // 灰阶
    grey: {
      50: '#fafafa',
      100: '#f5f5f5',
      200: '#eeeeee',
      300: '#e0e0e0',
      400: '#bdbdbd',
      500: '#9e9e9e',
      600: '#757575',
      700: '#616161',
      800: '#424242',
      900: '#212121',
      A100: '#f5f5f5',
      A200: '#eeeeee',
      A400: '#bdbdbd',
      A700: '#616161'
    },
    // 反色阈值，用于未设定字体色时自动装配最佳字体色
    contrastThreshold: 2.3,
    // 字体功能色
    text: {
      primary: 'rgba(0, 0, 0, 0.87)',
      secondary: 'rgba(0, 0, 0, 0.6)',
      disabled: 'rgba(0, 0, 0, 0.38)'
    },

    // 特殊色
    divider: 'rgba(0, 0, 0, 0.12)',
    background: {
      paper: '#ffffff',
      default: '#ffffff'
    },
    // 状态调色
    action: {
      active: 'rgba(0, 0, 0, 0.54)',
      hover: 'rgba(0, 0, 0, 0.04)',
      hoverOpacity: 0.04,
      selected: 'rgba(0, 0, 0, 0.08)',
      selectedOpacity: 0.08,
      disabled: 'rgba(0, 0, 0, 0.26)',
      disabledBackground: 'rgba(0, 0, 0, 0.12)',
      disabledOpacity: 0.38,
      focus: 'rgba(0, 0, 0, 0.12)',
      focusOpacity: 0.12,
      activatedOpacity: 0.12
    },
    // APIs,非特殊情况不要设置
    // 根据一个颜色计算反差色
    getContrastText: () => {},
    // 根据一个颜色名称计算反差色，是`getContrastText()`的包装，参数为预置色的名称
    getFixedContrastText: () => {},
    // 根据预置色名称获得ColorKeeper对象
    getColor: () => {},
    // 根据main色值计算其它必要色
    augmentColor: () => {}
  },

  typography: {
    // fontFamily: '',
    // 默认标准字号
    fontSize: 14,
    // 标准行高
    lineHeight: 1.2,
    // 微盟字重
    fontWeightBold: 700,
    fontWeightLight: 300,
    fontWeightMedium: 600,
    fontWeightRegular: 400,
    // 微盟字阶: t1 ~ t18
    textLevel: []
    // ...其它预置，key不可修改，可新增自己用
  },
  // 动画库
  animation: {
    // 内置方法, 根据 name, props从动画库中获取对应的css，现在只做了个spin😎
    create(name, props) {}
  },
  // 圆角半径
  radius: {
    // 内置方法，默认 num * 4px，获取
    create(num) {}
  },

  shadows: {
    // 内置方法, 内置25个3D阴影级别
    create(level) {}
  },
  // 根据默认4px *num 计算实际值,内置方法
  spacing() {},
  // 创建css transitions
  transitions: {
    create() {}
    // ... 其它详见
  },

  // 预置的功能性z-index，可以改、可以新增，但要注意顺序，不可删
  zIndex: {
    navBar: 1100,
    drawer: 1200,
    modal: 1300,
    pin: 1500
  },

  /**
   * @dev 全局统一使用的 props 值，在组件中使用 useThemeProps 时可通过参数将全局 Props 值作为默认 Props 引入
   * 例如：volume、color 等常用属性默认存在，可修改调整默认值，也可以新增新的值以供自己代码使用，但是需注意冲突问题
   */
  globalProps: {
    // 全局默认色调, 可选 primary,secondary,info,warning,error,success，可在调色盘中设置具体色值
    color: 'primary',
    // 综合大小（体积），可选 tiny,small,medium,large, 部分组件会使用到
    volume: 'medium'
  },
  // styled生成组件的配置项
  components: {
    // 内置组件前面加个Marrs
    MarrsButton: {
      // 默认props
      defaultProps: {
        textTransform: 'capitalize'
      },
      // 新增样式体，可以处理组件原本没有处理的入参
      variants: [
        {
          props: { face: 'dashed', color: 'primary' },
          style: {
            color: 'red',
            border: '1px dashed red'
          }
        }
      ],
      // 内置样式重写，忽略前缀
      styleOverrides: {
        dashedVolumeLarge: {
          // background: 'grey'
        }
      }
    }
  }
};

```

### 工具集

#### @wemo-ui/marrs/styles

##### useThemeProps

@param { props: inProps, name: ComponentName, customProps: {}, include: []}

##### useTheme

@return

usage

## 自定义样式

### 方式一: classes-slot，一般用于批量套用既有样式

```react
// 根据组件暴露出来的classes-slot, 可以插入自己的classname
<Button classes={{ root: 'my-classname', startIcon: 'my-classname-icon'}}>按钮</Button>

// 若不想写个css文件，更推荐使用jss方式生成样式，该方式可以使用主题参数
const useStyles = makeStyles((theme) => {
  return {
    root: {
      backgroundColor: theme.palette.getColor('info').lighten().color(),
      height: '500px'
    }
  };
});
// 然后在组件函数内部使用useStyles获得所有的classname
const classes = useStyles();
// 直接使用
<div className={classes.root}>^_^</div>
// 或者载入classes-slot
<Button classes={{ root: classes.root }}>按钮</Button>

```

### 方式二: sx 属性，一般用于微调

```react
// sx属性专门用于通过styled函数创建的 styled component，与直接写style的区别是sx可以使用主题
<Button sx={{ color: 'blue', padding: (theme) => theme.spacing(40) }}>
  超大自定义按钮
</Button>
```

### 方式三：修改 theme.components，新增样式实体，一般用于对组件库的改造

```react
// 修改theme.components下某个组件的variants属性
{
  variants: [{
    props: { face: 'dashed', color: 'primary' },
    style: {
      color: 'red',
      border: '1px dashed red'
    }
  }]
}
// 代码中可以直接使用扩展的属性样式
<Button face="dashed" color="primary">本来face属性没有dashed选项</Button>
```

### 方式四：使用 styled api 重建自定义组件，并继承原组件功能

```react
import { Button } from '@wemo-ui/marrs';
const MyButton = styled(Button)(({ theme }) => {
  background: theme.spacing(10)
});

<MyButton>我的按钮</MyButton>
```


CHANGELOG.md
# [0.1.0](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/compare/v1.1.21...v0.1.0) (2022-03-23)


### Bug Fixes

* **marrs-ui/dropdown:** dropdown组件功能完善样式调整 ([0c94df7](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/0c94df7edce14b058d561dc873d3edbf2ee0501b))
* **marrs-ui/popup:** 修复popup弹窗类组件居中问题 ([b86204e](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/b86204eaa34c3b88d690c5efae59867906d25ce0))
* **marrs-ui/popup:** 修复弹窗组件无法居中问题 ([1897b0c](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/1897b0ca40e4c90d9da2465be24f4c2f435b1c99))


### Features

* **marrs-ui/dropdown:** dropdown组件功能完善 ([45d2419](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/45d2419cf820be6a313c1acc91cdafbaf379836a))
* **marrs-ui/search、marrs-ui/field:** 调整searchBar居中样式 ([a662554](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/a66255466e96f2cc611f592e23ca6926b3c909a6))
* **marrs-ui/searchbar:** searbar组件居中样式调整 ([34c45d6](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/34c45d64d6f11463dee81df48d52ac9d67e69ae0))
* **marrs-ui/searchbar:** searchbar样式调整 ([8eac81d](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/8eac81da5bd3bb76aeb23f0dc811dffff2c41f5c))
* **marrs-ui/search:** 去除search组件无用代码 ([f39140d](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/f39140d7cd93864a5bddffd777abd916cc74ed9b))
* **marrs-ui/treeselect:** 增加sidebarWidth属性，当content为空时宽度只为sidebar宽 ([4b304b3](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/4b304b36a73222915bdfbee1c1bf04d2565334ab))
* **marrs-ui:** 优化组件ts注释，调整dropdown组件功能 ([32be61b](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/32be61b852ef9dcccd89c6e2abf3967c8df61c4a))



# [0.1.0](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/compare/v1.1.16...v0.1.0) (2022-02-18)


### Bug Fixes

* **marrs-ui/field:** field组件bug修复 ([a9fd95c](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/a9fd95c3f9a5cf16e82d4c55194e42c9e8836d8c))
* **marrs-ui/field:** 修复field组件bug ([9c8e555](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/9c8e5554e3fe3ead4baef73e8e971c04321e56b7))
* **marrs-ui/tab:** tab组件样式报错修复 ([a86fc6c](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits\/a86fc6c474c3e119a96e708d5dde67e7bde18cd8))
* **marrs-ui/tag:** 修改createBorder方法关于redius的问题 ([8103b4e](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/8103b4e5f152ce880807c25ecc4d7ef39c02c640))


### Features

* **marrs-ui/checkboxbase:** 调整medium规格大小 ([40a944e](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/40a944ea0ca66de3bc2563e183f6110741f9eebc))
* **marrs-ui/datetimepicker:** 去除dateTimepick组件对dayjs的依赖 ([5bad837](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/5bad8374f731a8b8116709774470c5bfeaa3c69d))
* **marrs-ui/field,searchbar:** field和searchBar组件增加autoFocus属性 ([e808bd1](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/e808bd155a3b2f5cb4525808053508a5f350077b))
* **marrs-ui/icon:** icon增加size属性，用调整icon大小 ([923d753](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/923d753722ab4a08d8078a87e5be884d2cca171e))
* **marrs-ui/icon:** 更新icon组件文档 ([4e08238](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/4e0823802f501df53a14f645e691f242fdc91db2))
* **marrs-ui/image,tag:** image组件增加圆角配置，tag组件调整圆角配置 ([92ea00e](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/92ea00e60ee4e5fd942b39dadc680ac307256d74))
* **marrs-ui/stepper:** stepper 组件文档更新 ([3fe6977](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/3fe697722b6007d3ae81547137e88022b9cf6532))
* **marrs-ui/stepper:** stepper组件增加加减号图标自定义配置 ([2ccde6d](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/2ccde6db59bb411a0da3327013d74a9f2afe5f34))
* **marrs-ui/tag:** tag组件样式调整 ([eb7a6a5](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/eb7a6a521f1df55562ca54bc566a680037017809))
* **mars-ui/field:** 调整field组件样式 ([384eff8](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/384eff8210c2f7de50b5ab0c370a1600233b09f6))



# [0.1.0](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/compare/v1.1.16...v0.1.0) (2022-01-18)


### Bug Fixes

* **marrs-ui/field:** 修复field组件bug ([9c8e555](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/9c8e5554e3fe3ead4baef73e8e971c04321e56b7))
* **marrs-ui/field:** field组件bug修复 ([a9fd95c](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/a9fd95c3f9a5cf16e82d4c55194e42c9e8836d8c))


### Features

* **marrs-ui/field,searchbar:** field和searchBar组件增加autoFocus属性 ([e808bd1](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/e808bd155a3b2f5cb4525808053508a5f350077b))
* **marrs-ui/icon:** 更新icon组件文档 ([4e08238](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/4e0823802f501df53a14f645e691f242fdc91db2))
* **marrs-ui/icon:** icon增加size属性，用调整icon大小 ([923d753](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/923d753722ab4a08d8078a87e5be884d2cca171e))
* **marrs-ui/stepper:** stepper 组件文档更新 ([3fe6977](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/3fe697722b6007d3ae81547137e88022b9cf6532))
* **marrs-ui/stepper:** stepper组件增加加减号图标自定义配置 ([2ccde6d](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/2ccde6db59bb411a0da3327013d74a9f2afe5f34))
* **marrs-ui/tag:** tag组件样式调整 ([eb7a6a5](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/eb7a6a521f1df55562ca54bc566a680037017809))



# [0.1.0](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/compare/v1.1.14...v0.1.0) (2022-01-12)


### Features

* **marrs-ui/field:** 修复Field组件number类型问题 ([f14d2de](https://stash.weimob.com/cloud-ability/fe-team/marrs-ui/commits/f14d2de1f0ea37cfafd7ec257e27a98920224757))

src/blackboard.js
/** 主题配置与API */
// const themeConfig = {
//   // 调色盘
//   palette: {
//     // [warning]以下内容必须存在，可以修改但不可置空，否则有可能会影响组件库的运行
//     common: {
//       black: '#000000',
//       white: '#ffffff'
//     },
//     // 快速模式切换, alpha
//     mode: 'main',

//     // 色调调色
//     primary: {
//       main: '#FA2C19'
//     },
//     secondary: {
//       main: '#9C27B0'
//     },
//     error: {
//       main: '#FF2E2E'
//     },
//     warning: {
//       main: '#FFA300',
//       // contrastText代表对应main为背景色时候的对照字体颜色，如果没有显示声明，默认补全为getContrastText(colorValue)
//       contrastText: 'rgba(0, 0, 0, 0.87)'
//     },
//     info: {
//       main: '#2A6AE9'
//     },
//     success: {
//       main: '#07C160'
//     },
//     // 中性色阶
//     neutral: {
//       N1: '#FFFFFF',
//       N2: '#FAFAFA',
//       N3: '#F5F5F5',
//       N4: '#F2F2F2',
//       N5: '#E0E0E0',
//       N6: '#C4C4C4',
//       N7: '#9E9E9E',
//       N8: '#757575',
//       N9: '#212121'
//     },
//     // 灰阶
//     grey: {
//       50: '#fafafa',
//       100: '#f5f5f5',
//       200: '#eeeeee',
//       300: '#e0e0e0',
//       400: '#bdbdbd',
//       500: '#9e9e9e',
//       600: '#757575',
//       700: '#616161',
//       800: '#424242',
//       900: '#212121',
//       A100: '#f5f5f5',
//       A200: '#eeeeee',
//       A400: '#bdbdbd',
//       A700: '#616161'
//     },
//     // 反色阈值，用于未设定字体色时自动装配最佳字体色
//     contrastThreshold: 2.3,
//     // 字体功能色
//     text: {
//       primary: 'rgba(0, 0, 0, 0.87)',
//       secondary: 'rgba(0, 0, 0, 0.6)',
//       disabled: 'rgba(0, 0, 0, 0.38)'
//     },

//     // 特殊色
//     divider: 'rgba(0, 0, 0, 0.12)',
//     background: {
//       paper: '#ffffff',
//       default: '#ffffff'
//     },
//     // 状态调色
//     action: {
//       active: 'rgba(0, 0, 0, 0.54)',
//       hover: 'rgba(0, 0, 0, 0.04)',
//       hoverOpacity: 0.04,
//       selected: 'rgba(0, 0, 0, 0.08)',
//       selectedOpacity: 0.08,
//       disabled: 'rgba(0, 0, 0, 0.26)',
//       disabledBackground: 'rgba(0, 0, 0, 0.12)',
//       disabledOpacity: 0.38,
//       focus: 'rgba(0, 0, 0, 0.12)',
//       focusOpacity: 0.12,
//       activatedOpacity: 0.12
//     },
//     // APIs,非特殊情况不要设置
//     // 根据一个颜色计算反差色
//     getContrastText: () => {},
//     // 根据一个颜色名称计算反差色，是`getContrastText()`的包装，参数为预置色的名称
//     getFixedContrastText: () => {},
//     // 根据预置色名称获得ColorKeeper对象
//     getColor: () => {},
//     // 根据main色值计算其它必要色
//     augmentColor: () => {}
//   },

//   typography: {
//     // fontFamily: '',
//     // 默认标准字号
//     fontSize: 14,
//     // 标准行高
//     lineHeight: 1.2,
//     // 微盟字重
//     fontWeightBold: 700,
//     fontWeightLight: 300,
//     fontWeightMedium: 600,
//     fontWeightRegular: 400,
//     // 微盟字阶: t1 ~ t18
//     textLevel: []
//     // ...其它预置，key不可修改，可新增自己用
//   },
//   // 动画库
//   animation: {
//     // 内置方法, 根据 name, props从动画库中获取对应的css，现在只做了个spin😎
//     create(name, props) {}
//   },
//   // 圆角半径
//   radius: {
//     // 内置方法，默认 num * 4px，获取
//     create(num) {}
//   },

//   shadows: {
//     // 内置方法, 内置25个3D阴影级别
//     create(level) {}
//   },
//   // 根据默认4px *num 计算实际值,内置方法
//   spacing() {},
//   // 创建css transitions
//   transitions: {
//     create() {}
//     // ... 其它详见
//   },

//   // 预置的功能性z-index，可以改、可以新增，但要注意顺序，不可删
//   zIndex: {
//     navBar: 1100,
//     drawer: 1200,
//     modal: 1300,
//     pin: 1500
//   },

//   /**
//    * 全局统一使用的 props 值，在组件中使用 useThemeProps 时可通过参数将全局 Props 值作为默认 Props 引入
//    * 例如：volume、color 等常用属性默认存在，可修改调整默认值，也可以新增新的值以供自己代码使用，但是需注意冲突问题
//    */
//   globalProps: {
//     // 全局默认色调, 可选 primary,secondary,info,warning,error,success，可在调色盘中设置具体色值
//     color: 'primary',
//     // 综合大小（体积），可选 tiny,small,medium,large, 部分组件会使用到
//     volume: 'medium'
//   },
//   // styled生成组件的配置项
//   components: {
//     // 内置组件前面加个Marrs
//     MarrsButton: {
//       // 默认props
//       defaultProps: {
//         textTransform: 'capitalize'
//       },
//       // 新增样式体，可以处理组件原本没有处理的入参
//       variants: [
//         {
//           props: { face: 'dashed', color: 'primary' },
//           style: {
//             color: 'red',
//             border: '1px dashed red'
//           }
//         }
//       ],
//       // 内置样式重写，忽略前缀
//       styleOverrides: {
//         dashedVolumeLarge: {
//           // background: 'grey'
//         }
//       }
//     }
//   }
// };
